src/tiled/commanddialog.cpp

```diff
@@ -76,19 +84,55 @@ void CommandDialog::setShortcut(const QKeySequence &keySequence)
         mUi->treeView->model()->setShortcut(current, keySequence);
 }
 
-void CommandDialog::updateKeySequenceEdit(const QModelIndex &current, const QModelIndex &)
+void CommandDialog::setSaveBeforeExecute(int state)
+{
+    const QModelIndex &current = mUi->treeView->currentIndex();
+    if (current.row() < mUi->treeView->model()->rowCount(QModelIndex()))
+        mUi->treeView->model()->setSaveBeforeExecute(current, state);
+}
+
+void CommandDialog::setCommand(const QString &text)
+{
+    const QModelIndex &current = mUi->treeView->currentIndex();
+    if (current.row() < mUi->treeView->model()->rowCount(QModelIndex()))
+        mUi->treeView->model()->setCommand(current, text);
+}
+
+void CommandDialog::updateWidgets(const QModelIndex &current, const QModelIndex &)
 {
     if (current.row() < mUi->treeView->model()->rowCount(QModelIndex()) - 1) {
+        mUi->commandEdit->setEnabled(true);
+        mUi->browseButton->setEnabled(true);
         mUi->keySequenceEdit->setEnabled(true);
         mUi->clearButton->setEnabled(true);
+        mUi->saveBox->setEnabled(true);
+
         mUi->keySequenceEdit->setKeySequence(mUi->treeView->model()->shortcut(current));
-    } else {
+        mUi->saveBox->setChecked(mUi->treeView->model()->saveBeforeExecute(current));
+        mUi->commandEdit->setText(mUi->treeView->model()->command(current));
+    }
+    else {
+        mUi->commandEdit->clear();
         mUi->keySequenceEdit->clear();
+        mUi->saveBox->setChecked(false);
+        mUi->commandEdit->setEnabled(false);
+        mUi->browseButton->setEnabled(false);
         mUi->keySequenceEdit->setEnabled(false);
         mUi->clearButton->setEnabled(false);
+        mUi->saveBox->setEnabled(false);
     }
 }
 
+void CommandDialog::openFileDialog()
+{
+    QString caption = tr("Select Executable");
+    QString dir = QStandardPaths::writableLocation(QStandardPaths::ApplicationsLocation);
+    QString executableName = QFileDialog::getOpenFileName(this, caption, dir);
+
+    if (!executableName.isEmpty())
+        mUi->commandEdit->setText(executableName);
```

Comment:
I think if we don't always add it here, we need to introduce a separate "arguments" field, like in Qt Creator. That way the arguments can default to `%mapfile` and choosing the executable becomes independent.